marklogic-data-hub/src/main/resources/ml-modules/root/data-hub/5/builtins/steps/mapping/entity-services/lib.sjs

```diff
@@ -141,16 +167,34 @@ function buildMapProperties(mapping, entityModel) {
   return propertyLines.join('\n');
 }
 
-function getRelatedMappings(mapping, related = [mapping]) {
-  // get references to sub mappings
+/**
+ * Recursive function that returns a mapping for each property with a targetEntityType, which signifies that it is
+ * mapping to an object property. Each of these will need to be converted into an m:entity XML template. The name of
+ * each template is guaranteed to be unique by being based on the propertyPath and the title of each object property
+ * being mapped. This ensures that we have uniquely-named templates in the XSLT transform that's generated from the
+ * XML mapping template.
+ *
+ * @param mapping
+ * @param propertyPath
+ * @param objectPropertyMappings
+ * @return {*[]}
+ */
+function getObjectPropertyMappings(mapping, propertyPath, objectPropertyMappings = []) {
   if (dhMappingTraceIsEnabled) {
     xdmp.trace(dhMappingTrace, `Getting related mappings for '${xdmp.describe(mapping)}'`);
   }
-  let internalMappings = mapping.xpath('/properties//object-node()[exists(targetEntityType) and exists(properties)]');
-  for (let internalMapping of internalMappings) {
-    related.push(internalMapping);
-  }
-  return related;
+  Object.keys(mapping.properties).forEach(propertyTitle => {
+    const property = mapping.properties[propertyTitle];
```

Comment:
We have a failing test. null check required here.

`com.marklogic.client.FailedRequestException: Local message: failed to apply resource at documents: Internal Server Error. Server Message: JS-JAVASCRIPT: Object.keys(mapping.properties).forEach(propertyTitle => { -- Error running JavaScript request: TypeError: Cannot convert undefined or null to object . See the MarkLogic server error log for further detail.`

Reply:
I'll fix it, but it's really due to an invalid test file - that test is generating a mapping file with "null" as the value of "properties", which doesn't make any sense.

Reply:
True that about the mapping file in the test. When a mapping step is created a version 0 file has similar structure ie no properties.